rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,400 @@
+
+**Request:**
+| Parameter | Description |
+| -------------- | --------------------------------------------------------------------------------------------- |
+| dataset_name | Identifies the dataset the gene sets are linked to. |
+| user_id | Identifies the user creating the gene set |  
+| body | JSON formatted dict containing a list of gene sets (and accompanying genes) to be created |
+
+```json
+{
+  "gene_sets": [
+    {
+      "gene_set_name": "[Required] gene set name",
+      "comments": "[OPTIONAL]Comments describing gene set",
+      "genes": [
+        {
+          "gene name": "[Required] name of gene",
+          "comments": "[OPTIONAL] comments on why a gene was included in this gene set"
+        }
+      ]
+    }
+  ]
+}
+```
+
+**Response:**
+| Code | Description |
+| ---- | --------------------------------------- |
+| 201 | The gene set(s) were successfully saved |
+
+| Key      | Description                                                                               |
+| -------- | ----------------------------------------------------------------------------------------- |
+| gene_set | Dict mapping the gene set name to its assigned uuid and its consensus counter             |
+| errors   | In the case where some gene sets were successfully stored and others will not, a message  |
+|          | specifying the error and the gene set(s) it was linked to is returned to the client here. |
+
+```json
+{ "name_of_gene_set": { "uuid": "gene_set_uuid", "consensus_counter": "0001" } }
+```
+
+**Error Responses:**
+
+| Code | Description                             |
+| ---- | --------------------------------------- |
+| 400  | If the parameters supplied are invalid. |
+
+| Code | Description                                                  |
+| ---- | ------------------------------------------------------------ |
+| 403  | User id invalid, or user does not have access to the dataset |
+
+#### PUT dataset_name/api/v0.1/gene_sets/{gene_set_uuid}
+
+Any changes to a gene set (the gene set itself or genes within that gene set) will be sent via json to the backend.
+This includes the creation/deletion of genes within a gene set
+Updates to comments will overwrite the currently stored comments
+**Request:**
+
+| Parameter     | Description                                                                                  |
+| ------------- | -------------------------------------------------------------------------------------------- |
+| dataset_name  | Identifies the dataset the gene sets are linked to.                                          |
+| user_id       | Identifies the user creating the gene set                                                    |
+| gene_set_uuid | UUID of gene set to be updated                                                               |
+| body          | JSON formatted dict containing a gene set (and accompanying genes) to be updated-- see below |
+
+```json
```

Comment:
This interface adds some complexity to the client.  The client will need to have accurate knowledge of the state of the database, and the new state it wants for the gene set, in order to produce the add/delete/update lists.    The interface is not idempotent, so retrying after a failure may fail.   So if a failure does happen, the client has to do a POST.

Reply:
Yes. @mckinsel suggested having the client own the counter and when a request with the wrong counter number came in then the server could tell the client what the latest update it had processed was and the client could either resend the next the updates or send a post request recreating the gene set